lib/kubernetes/src/equality/ServiceMonitor.ts

```diff
@@ -14,63 +14,22 @@
  * limitations under the License.
  */
 
-import { isDeepStrictEqual } from "util";
+import equal from "fast-deep-equal";
 import { V1Servicemonitor } from "..";
 
 export const isServiceMonitorEqual = (
   desired: V1Servicemonitor,
   existing: V1Servicemonitor
 ): boolean => {
-  if (!isDeepStrictEqual(desired.spec.selector, existing.spec.selector)) {
+  if (!equal(desired.spec.selector, existing.spec.selector)) {
     return false;
   }
   if (desired.spec.jobLabel !== existing.spec.jobLabel) {
     return false;
   }
-  if (!areEndpointsEqual(desired, existing)) {
+  if (!equal(desired, existing)) {
```

Comment:
```suggestion
  if (!equal(desired.spec.endpoints, existing.spec.endpoints)) {
```

Otherwise, the controller will always try to update this resource because the equality check always fails. The check doesn't account for the fields Kubernetes sets in the metadata when the resource is created. 

Here's an example of how the resource looks like when it's created in Kubernetes.
```
> k get servicemonitors.monitoring.coreos.com -n system-tenant prometheus -o yaml
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  annotations:
    opstrace: owned
  creationTimestamp: "2021-05-07T11:19:46Z"
  generation: 1
  labels:
    app: prometheus
    prometheus: system-prometheus
    tenant: system
  name: prometheus
  namespace: system-tenant
  resourceVersion: "2795"
  selfLink: /apis/monitoring.coreos.com/v1/namespaces/system-tenant/servicemonitors/prometheus
  uid: 9d6a8309-2584-49e9-8090-329e0715cc0d
spec:
  endpoints:
  - interval: 30s
    path: /prometheus/metrics
    port: web
  jobLabel: prometheus
  selector:
    matchLabels:
      app: prometheus
```

Reply:
Good catch!

Reply:
Committed your suggestion, and also [added a test](https://github.com/opstrace/opstrace/pull/685/files#diff-b639d9f58c083bd9922e86c84af6a4577d0dbdf2c0c8e6b0f2f0024dd51d5abdR72-R81) for it